### PR TITLE
Minikube Blog Post

### PR DESCRIPTION
**Pull request by JoshArmi**

### CODE REVIEW

File: _posts/2018-08-03-bashing-through-minikube.markdown [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+### 1. Every project in a repository should be testable in isolation
+With Docker it is so trivial to spin up a database container that people can make the mistake of making all their component tests require an actual database to function. This complicates matters significantly especially when you add parallel docker environments between minikube and your local machine.
+
+### 2. Don't rely on on being able to run your entire stack locally
+When you have a small number of services it is possible to run them all locally at once, however this stops being the case quickly, and is more difficult to retroactively fix then to stop this at the start.
+
+The need to do this is generally driven by insufficient testing, or testing at the wrong isolation levels. For testing across repositories leave that to Continuous Delivery.
+
+
+## Wrap Up
```
Comment (jgunnink) [inline _posts/2018-08-03-bashing-through-minikube.markdown]: your previous headings throughout the article are `Sentence case`. This heading is `Title Case`. Up to you which you want to use, but you should pick one and make it consistent through the article.

File: _posts/2018-08-03-bashing-through-minikube.markdown [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+###  1. Script your Local Build and Local Run
+With microservices you can reduce the context switch penalty by providing a common interface for every repository. I've found having the following 5 scripts in every repository sames time and many headaches:
+
+1. `build.sh`
+    This script builds the artifacts, i.e. this is where you call `mvn clean verify` or `npm install`
+    If you have a lot of artifacts generated from a repository it can be useful to make it so you can pass an artifact name to the script and build that in isolation for quicker feedback.
+2. `local-build.sh`
+    This script idempotently sets up the local environment, i.e. make sure minikube is running, make sure minikube has required dependent resources configured.
+3. `local-run.sh`
+    This script deploys Kubernetes manifests to Minikube, either via Helm, Kubectl or some other tooling. Also you can run integration tests here.
```
[the remark this answers is not in the document]
Comment (JoshArmi) [reply _posts/2018-08-03-bashing-through-minikube.markdown]: I'll link for Helm, but if they don't know what Kubectl is then I'm not sure what value they'll get from the post as a whole

File: _posts/2018-08-03-bashing-through-minikube.markdown [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+#### 4. Kubernetes doesn't come fast or easily
+The current thinking is becoming proficient with Kubernetes is a 3 to 6 month journey. If you're not planning on running Kubernetes in production you can get simple local container orchestration with `docker compose`.
+
+
+## Patterns
+###  1. Script your Local Build and Local Run
+With microservices you can reduce the context switch penalty by providing a common interface for every repository. I've found having the following 5 scripts in every repository sames time and many headaches:
+
+1. `build.sh`
+    This script builds the artifacts, i.e. this is where you call `mvn clean verify` or `npm install`
```
[the remark this answers is not in the document]
Comment (JoshArmi) [reply _posts/2018-08-03-bashing-through-minikube.markdown]: Yes 😛

File: _posts/2018-08-06-bashing-through-minikube.markdown [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+## Minikube You Say?
+When you begin on a Kubernetes journey, often one of the first questions is how do I run locally? That's where Minikube comes in, you get an easy to spin up single node Kubernetes environment where you can rapidly iterate. As with nearly all tools, there's some rough edges and gotchas to figure out, in the next 5 minutes I hope to save you hours of frustration.
+
+The blog was written using MacOS but the commands should translate across Linux distributions.
+
+
+## Gotchas
+#### 1. What `eval $(minikube docker-env)` Does
+When you run Minikube it's running in a virtual machine, and in effect you are now running two parallel *docker environments*, the one on your local machine and the one in the virtual machine. The `eval` command does is set a selection of environment variables such that your current terminal session is pointing to the *docker environment* in the virtual machine. This means all the images you have locally won't be available and vice versa.
```
Comment (Hamo69) [inline _posts/2018-08-06-bashing-through-minikube.markdown]: *What* the `eval` command does..

File: _posts/2018-08-06-bashing-through-minikube.markdown [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+layout: post
+title:  "Bashing through Minikube - 0 to 60 for Local Kubernetes Development"
+date:   2018-08-06
+tags: kubernetes minikube containers bash local development
+author: Josh Armitage
+image: img/kubernetes/minikube-logo.png
+---
+
+## Minikube You Say?
+When you begin on a Kubernetes journey, often one of the first questions is how do I run locally? That's where Minikube comes in, you get an easy to spin up single node Kubernetes environment where you can rapidly iterate. As with nearly all tools, there's some rough edges and gotchas to figure out, in the next 5 minutes I hope to save you hours of frustration.
```
Comment (Hamo69) [inline _posts/2018-08-06-bashing-through-minikube.markdown]: When you begin on a Kubernetes journey, often one of the first questions is; "how do I run locally"?

Comment (Hamo69) [inline _posts/2018-08-06-bashing-through-minikube.markdown]: That's where Minikube comes in; you get an easy to spin up single node Kubernetes environment which allows you to rapidly iterate.

Comment (Hamo69) [reply _posts/2018-08-06-bashing-through-minikube.markdown]: potentially split the last sentence of the 1st para into two sentences.

File: _posts/2018-08-06-bashing-through-minikube.markdown [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+This leads into the core of your continuous integration configuration being `build.sh && publish.sh`
+
+### 2. Keep Your Containers Small
+Using multi-stage Docker builds, i.e one container to build and one minimal container for runtime, will keep your feedback cycles fast. For example, with Java your `build` container can have the JDK installed with all your testing dependencies, but then you can create a JRE only Alpine container that can be about 5-10x smaller.
+
+<div style="text-align: center; margin: 10px"><iframe width="560" height="315" src="https://www.youtube.com/embed/wGz_cbtCiEA?rel=0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe></div>
+
+### 3. Have Your Container Entrypoint And Command Be Targeted
+The containers that you are expecting to run in your cluster should be single purpose by design, that means that you can have your entrypoint and command set so you don't have to manage that outside the Dockerfile itself.
```
Comment (Hamo69) [inline _posts/2018-08-06-bashing-through-minikube.markdown]: a couple too many "that"s in this sentence

File: _posts/2018-08-06-bashing-through-minikube.markdown [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+### 2. Don't Rely On Being Able To Run Your Entire Stack Locally
+When you have a small number of services it is possible to run them all locally at once, however this stops being the case quickly and is more difficult to retroactively fix then to stop this at the start.
+
+The need to do this is generally driven by insufficient testing, or testing at the wrong isolation levels. For testing across repositories leave that to Continuous Delivery.
+
+## Wrap Up
+Hopefully this post will have saved you many frustrations in the road ahead to Kubernetes adoption. Minikube is fantastic for keeping your feedback cycles fast and learning its quirks is well worth the time investment.
+
+The next step is getting to grips with your Kubernetes flavour of choice, be that cloud vendor managed or for those stout of heart a bare metal installation. Stay tuned for posts on adopting EKS, up-skilling the team into Kubernetes, and making your cluster production ready.
```
Comment (Hamo69) [inline _posts/2018-08-06-bashing-through-minikube.markdown]: can you provide a link to EKS?

Stay tuned for posts on adopting [EKS](https://aws.amazon.com/eks/)

File: _posts/2018-08-06-bashing-through-minikube.markdown [hunk 8]
```diff
@@ -0,0 +1,91 @@
+---
+layout: post
+title:  "Bashing through Minikube - 0 to 60 for Local Kubernetes Development"
+date:   2018-08-06
```
Comment (Hamo69) [inline _posts/2018-08-06-bashing-through-minikube.markdown]: maybe set the date for Wednesday ie. 2018-08-08

**Comment by Hamo69** [comment]
yo @JoshArmi can you amend the date to 2018-08-13 and merge prior to that once comments are addressed? We can push it out via socials etc on Monday then

File: _posts/2018-08-06-bashing-through-minikube.markdown (hunk 8 above)
Comment (temyers) [reply _posts/2018-08-06-bashing-through-minikube.markdown]: I think next Monday would be good - means there's a good flow of posts, but not all at once.

File: _posts/2018-08-06-bashing-through-minikube.markdown [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+When you begin on a Kubernetes journey, often one of the first questions is how do I run locally? That's where Minikube comes in, you get an easy to spin up single node Kubernetes environment where you can rapidly iterate. As with nearly all tools, there's some rough edges and gotchas to figure out, in the next 5 minutes I hope to save you hours of frustration.
+
+The blog was written using MacOS but the commands should translate across Linux distributions.
+
+
+## Gotchas
+#### 1. What `eval $(minikube docker-env)` Does
+When you run Minikube it's running in a virtual machine, and in effect you are now running two parallel *docker environments*, the one on your local machine and the one in the virtual machine. The `eval` command does is set a selection of environment variables such that your current terminal session is pointing to the *docker environment* in the virtual machine. This means all the images you have locally won't be available and vice versa.
+#### 2. Changing Back To Your Local Docker Environment
+The magic rune to swap you back to your local docker environment is:
```
Comment (temyers) [inline _posts/2018-08-06-bashing-through-minikube.markdown]: or open a new terminal? :)

File: _posts/2018-08-06-bashing-through-minikube.markdown [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+`unset DOCKER_API_VERSION DOCKER_TLS_VERIFY DOCKER_CERT_PATH DOCKER_API_VERSION DOCKER_HOST`
+#### 3. Swapping Between Minikube And Other Clusters
+Cluster autentication is managed via `config` files in `~/.kube/`. When you run `minikube start` it overwrites `~/.kube/config` by default, at the moment there is no command I have found for regenerating the config apart from starting the cluster again.
+#### 4. Kubernetes Doesn't Come Fast Or Easily
+The current thinking is becoming proficient with Kubernetes is a 3 to 6 month journey. If you're not planning on running Kubernetes in production you can get simple local container orchestration with `docker compose`.
+
+## Patterns
+###  1. Script Your Local Build And Local Run
```
Comment (temyers) [inline _posts/2018-08-06-bashing-through-minikube.markdown]: Can you link to examples for these scripts?

File: _posts/2018-08-06-bashing-through-minikube.markdown [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    This script deploys Kubernetes manifests to Minikube, either via Helm, Kubectl or some other tooling. Also you can run integration tests here.
+4. `local-nuke.sh`
+    This script remove the resources from Kubernetes and deletes the Docker image artifact
+5. `publish.sh`
+    This script takes your artifacts and publishes them to your artifact repository.
+
+This leads into the core of your continuous integration configuration being `build.sh && publish.sh`
+
+### 2. Keep Your Containers Small
+Using multi-stage Docker builds, i.e one container to build and one minimal container for runtime, will keep your feedback cycles fast. For example, with Java your `build` container can have the JDK installed with all your testing dependencies, but then you can create a JRE only Alpine container that can be about 5-10x smaller.
```
Comment (temyers) [inline _posts/2018-08-06-bashing-through-minikube.markdown]: Add a link for _multi-stage Docker builds_ to the appropriate doco page

**Comment by Hamo69** [comment]
@JoshArmi how are we going with revisions?

File: _posts/2018-08-06-better-bash-in-5-minutes.markdown [hunk 12]
```diff
@@ -1,7 +1,11 @@
 ---
 layout: post
 title:  "Better Bash Scripting in 5 Minutes or Less"
+<<<<<<< HEAD
 date:   2018-08-06
+=======
+date:   2018-08-10
+>>>>>>> 652d755... Changed categories to tags
```
Comment (jgunnink) [inline _posts/2018-08-06-better-bash-in-5-minutes.markdown]: :(

**Pull request merged**